### ISSUE TITLE
intro-2

### ISSUE THREAD

**Issue by 3v0k4** [AkkarinForest/Haskell-Workshop#3]
- [x] Try all snippets in ghci

**Comment by alex-suminski** [comment]
About multiline - I added in the readme that users should have `:set +m` in the `.ghci`. This makes it easy to create multiline code, without `:{`. Maybe that would be enough in your case as well?

**Comment by alex-suminski** [comment]
My suggestions, feel free to disagree:

- [x] 1. Use lambdas only if necessary. Eg this can be written with normal Ruby methods:
```
transform_first = ->(xs) {  xs[0].upcase }

transform_first.call(["hello"])
-> "HELLO"

transform_first.call([])
-> BOOM
```
Justification: Using lambdas in ruby examples adds complexity and probably some participants won't be familiar with it. I'd suggest we focus on teaching them Haskell and keep Ruby simple as it should be part of an explanation and new material.

- [x]  2. Mark which examples should be executed in the ghci and which if any require a file.

- [x]  3. If a student should use an example in the ghci it should work after copy-pasting. Eg I can not execute this:
```
safeHead :: [a] -> Maybe a
safeHead [] = Nothing
safeHead x:xs = Just x
```
cause type definitions don't work in ghci (unless there is some option I can enable?), `x:xs` in missing parentheses, and there needs to be a let:
```
Prelude> let safeHead [] = Nothing
Prelude|     safeHead (x:xs) = Just x
```

- [x]  4. If some examples require creating and compiling a file, add instruction in the main readme.

- [x]  5. As a proposed solution for parse exercises, this would be more readable:
```
parse xs = fmap (\string -> (upper string) ++ "!") (safeHead xs)
```
the version with reverse makes more sense if you want to use `.` or `$`, but I'm not sure if it's necessary here.

- [x]  6. bool values should be capitalized:
```
boolToString bool =
    case bool of
        True -> "True!!"
        False -> "False :("
```

**Comment by 3v0k4** [comment]
> Justification: Using lambdas in ruby examples adds complexity and probably some participants won't be familiar with it. I'd suggest we focus on teaching them Haskell and keep Ruby simple as it should be part of an explanation and new material.

I guess the alternative to using lambdas is to do something like

```ruby
def fnc(x)
  x+1
end

method(:fnc).call(123)
```

To me that is not ideal for couple of reasons:
- `method` is a Ruby only thing; cannot reuse that knowledge anywhere else.
- Lambdas (i.e. anonymous functions) nowadays you can find in any language; Ruby has a strange syntax though.
- It's way more similar to what somebody would use in a more functional setting
  ```
  hs = \x -> x + 1
  rb = ->(x) { x + 1 }
  ```
- With lambas in Ruby there's more freedom in where to define them or even if naming them.

But I'm a real biased person, I'll post in Lunar Slack and see what people grasp better between

```rb
plus_one = ->(x) { x + 1 }
plus_one.call(2)
# => 3
```

and

```rb
def plus_one(x)
  x + 1
end
method(:plus_one).call(1)
# => 2
```

I'll report back asap.

> bool values should be capitalized

Argh, in PureScript they are not. I'm always doing it wrong!

Thanks for reviewing, I'll fix stuff and let you know here.

**Comment by alex-suminski** [comment]
I don't know where and why you need the `x+1` code snippets you mention, but I meant that you don't need lambdas in beginning of Maybe chapter. What if you just had a an example like this:
```
>> def transformFirst(xs)
>>   xs.first.upcase
>> end
=> :transformFirst
>> transformFirst(["hello"])
=> "HELLO"
>> transformFirst([])
NoMethodError: undefined method `upcase' for nil:NilClass
```

**Comment by 3v0k4** [comment]
I wasn't being specific about intro-2, just talking in general. In the specific case for `transform_first` it's not needed. I think we used to do `method(:and_more)` in intro-1 before moving to lambdas. Since now we have lambdas there I think it could be fine to use them here too.

I got taken into some other things today. Not sure if I'll manage to go through the comments before tomorrow, sorry.

**Comment by 3v0k4** [comment]
Fixed (almost) all. I'm leaving the Ruby lambdas for now since they are well explained in intro-1. To be honest the more I look at those lambdas the more I want to start using them at work. Not joking. They look so much better than that `def` blob. I would even consider not using `def`s anymore after we explain it in intro-1. But I'm fine with keeping them too 👍 

I wonder what people would say if I went 100% Haskell and used lambdas and [`>>`](https://ruby-doc.org/core-2.7.1/Proc.html#method-i-3E-3E) 😄
